File: Test/AdventureWorksFunctionalModel/Human Resources/Shift.cs
namespace AW.Types;

[Bounded]
public class Shift : IHasModifiedDate
{
    public Shift() { }

    public Shift(Shift cloneFrom)
    {
        ShiftID = cloneFrom.ShiftID;
        Name = cloneFrom.Name;
        StartTime = cloneFrom.StartTime;
        EndTime = cloneFrom.EndTime;
        ModifiedDate = cloneFrom.ModifiedDate;
    }

    [Hidden]
    public byte ShiftID { get; init; }

    [MemberOrder(1)]
    public string Name { get; init; } = "";

    [MemberOrder(3), Mask("T")]
    public TimeSpan StartTime { get; init; }

    [MemberOrder(4), Mask("T")]
    public TimeSpan EndTime { get; init; }


    [MemberOrder(99), Versioned]
    public DateTime ModifiedDate { get; init; }

    public override string ToString() => Name;
}

File: Test/AdventureWorksLegacyModelVB/Human Resources/Shift.vb
Namespace AW.Types

	'<Bounded>
	Partial Public Class Shift

		Public Property ShiftID() As Byte

#Region "Name"
		Friend mappedName As String
		Friend myName As TextString

		'<MemberOrder(1)>
		Public ReadOnly Property Name As TextString
			Get
				Return If(myName, New TextString(mappedName, Function(v) mappedName = v))
			End Get
		End Property

		Public Sub AboutName(a As FieldAbout, Name As TextString)
			Select Case a.TypeCode
				Case AboutTypeCodes.Name
				Case AboutTypeCodes.Usable
				Case AboutTypeCodes.Valid
				Case AboutTypeCodes.Visible
			End Select
		End Sub
#End Region

		'<MemberOrder(3), Mask("T")>
		Public Property StartTime() As TimeSpan 'TODO

		'<MemberOrder(4), Mask("T")>
		Public Property EndTime() As TimeSpan 'TODO

#Region "ModifiedDate"
		Friend mappedModifiedDate As Date
		Friend myModifiedDate As TimeStamp

		'<MemberOrder(99)>
		Public ReadOnly Property ModifiedDate As TimeStamp
			Get
				Return If(myModifiedDate, New TimeStamp(mappedModifiedDate, Function(v) mappedModifiedDate = v))
			End Get
		End Property

		Public Sub AboutModifiedDate(a As FieldAbout, ModifiedDate As TimeStamp)
			Select Case a.TypeCode
				Case AboutTypeCodes.Usable
					a.Usable = False
			End Select
		End Sub
#End Region

		Public Function Title() As Title
			Return New Title(Name)
		End Function
	End Class
End Namespace